src/app/shared/pagination/pagination.directive.ts

```diff
@@ -0,0 +1,137 @@
+import {
+    Directive,
+    Input,
+    OnChanges,
+    OnInit,
+    SimpleChanges,
+    TemplateRef,
+    ViewContainerRef,
+    inject,
+} from '@angular/core';
+import {BehaviorSubject, filter, map} from 'rxjs';
+import {PaginationContext} from './pagination-context.interface';
+
+@Directive({
+    selector: '[appPagination]',
+})
+export class PaginationDirective<T> implements OnInit, OnChanges {
+    private readonly currentIndex$ = new BehaviorSubject<number>(0);
+
+    private readonly templateRef = inject<TemplateRef<PaginationContext<T>>>(TemplateRef);
+    private readonly viewContainerRef = inject(ViewContainerRef);
+
+    private chunkSize: number | null = null;
+
+    @Input() appPaginationOf: T[] | null | undefined;
+    @Input() set appPaginationChunkSize(cs: number) {
+        this.chunkSize = cs;
+    }
+
+    //  eslint-disable-next-line  @typescript-eslint/naming-convention
+    static ngTemplateGuard_appPaginationOf<T>(
+        _directive: PaginationDirective<T>,
+        _inputValue: unknown,
+    ): _inputValue is T[] {
+        return true;
+    }
+
+    static ngTemplateContextGuard<T>(
+        _directive: PaginationDirective<T>,
+        _context: unknown,
+    ): _context is PaginationContext<T> {
+        return true;
+    }
+
+    ngOnChanges({appPaginationOf}: SimpleChanges): void {
+        if (appPaginationOf) {
+            this.updateView();
+        }
+    }
+
+    ngOnInit(): void {
+        this.listenCurrentIndex();
+    }
+
+    private updateView() {
+        if (this.shouldShowItems()) {
+            this.currentIndex$.next(0);
+
+            return;
+        }
+
+        this.viewContainerRef.clear();
+    }
+
+    private listenCurrentIndex() {
+        this.currentIndex$
+            .pipe(
+                filter(() => this.shouldShowItems()),
+                map(currentIndex => this.getCurrentContext(currentIndex)),
+            )
+            .subscribe(context => {
+                this.viewContainerRef.clear();
+                this.viewContainerRef.createEmbeddedView(this.templateRef, context);
+            });
+    }
+
+    private shouldShowItems(): boolean {
+        return Boolean(this.appPaginationOf?.length);
+    }
+
+    private getCurrentContext(currentIndex: number): PaginationContext<T> {
+        const appPaginationOf = this.appPaginationOf as T[];
+
+        return {
+            $implicit: this.getProductsByPage(currentIndex),
+            appPaginationOf,
+            index: currentIndex,
+            pageIndexes: [...new Array(this.getPageCount()).keys()],
```

Comment:
Рекомендация: выглядит накрученно честно говоря: создали новый массив нужной длинны, достали его ключи и на основе их создали новый массив. В таких кейсах частно используют подобную запись: создают новый массивы нужной длинны и указывают функцию для генерации значения массива при итерациях по объекту переданному в качестве первого аргумента
```suggestion
            pageIndexes: Array.from({length: this.getPageCount()}, (_, index) => index),
```